core/trigonometry.cpp

```diff
@@ -11,56 +11,161 @@
 
 namespace scipp::core {
 
-Variable to_rad(Variable var) {
-  const auto scale = makeVariable<double>(Dims(), Shape(),
-                                          units::Unit(units::rad / units::deg),
-                                          Values{pi<double> / 180.0});
-  return var * scale;
+const auto deg_to_rad =
+    makeVariable<double>(Dims(), Shape(), units::Unit(units::rad / units::deg),
+                         Values{pi<double> / 180.0});
+
+Variable sin(const VariableConstProxy &var) {
+  using std::sin;
+  Variable out(var);
+  sin(out, out);
+  return out;
 }
 
-Variable sin(const Variable &var) {
+Variable sin(Variable &&var) {
   using std::sin;
-  const Variable &rad = var.unit() == units::deg ? to_rad(var) : var;
-  return transform<double, float>(
-      rad, overloaded{transform_flags::expect_no_variance_arg<0>,
-                      [](const auto &x) { return sin(x); }});
+  Variable out(std::move(var));
+  sin(out, out);
+  return out;
+}
+
+VariableProxy sin(const VariableConstProxy &var, const VariableProxy &out) {
+  using std::sin;
+  out.assign(var);
+  if (var.unit() == units::deg)
+    out *= deg_to_rad;
+  transform_in_place<pair_self_t<double, float>>(
+      out, out,
+      overloaded{transform_flags::expect_no_variance_arg<0>,
+                 transform_flags::expect_no_variance_arg<1>,
+                 [](auto &x, const auto &y) { x = sin(y); }});
+  return out;
```

Comment:
This ends up iterating over the array up to *3* times. The way it is written you also could have avoided the `transform` with 2 arguments and just use 1 arg instead.

That being said, for `out=in` with input in `rad` there is only a single iteration, so not everything is inefficient. I suggest you just switch back to a single-arg transform and leave the remaining possible "optimizations" until we know it is performance critical.

Reply:
> This ends up iterating over the array up to 3 times.

I don't think there is a nice way around this if you want to keep the unit conversion.
I did try and move the conversion into the transform but Boost units makes this difficult.

Reply:
Even so, the worst theoretical case (non in place with input in degress) is still faster then it originally was with these changes.

Reply:
Yes, as I said: Leave it, just fix the `transform_in_place`.

Reply:
(I see you did, github notifications still do not work with force-pushes, apparently)